src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+        action.keysToRemove.forEach((key) => delete currentState[key]);
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    stateHistory.push(currentState);
```

Comment:
The current implementation pushes `currentState` directly into `stateHistory`. This could lead to issues because objects in JavaScript are referenced, not copied. You should push a deep clone of `currentState` to ensure that each entry in `stateHistory` is independent of changes made in subsequent actions. Consider using a method like `JSON.parse(JSON.stringify(currentState))` to create a deep copy.